### ISSUE TITLE
Resource indicators are missing in access token request

### ISSUE THREAD

**Issue by pago93** [dotnet/aspnetcore#41176]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

The Resource-property set in the OpenIdConnectOptions does not flow through the whole process.

When configuring OpenIdConnect authentication with a resource indication, the resource value is only added to the initial authorization request but not to the subsequent access token request.

### Expected Behavior

When setting the Resource-property the value should be set in subsequent requests and no additional configuration should be needed.
As per [RFC 8707 Section 2.2](https://datatracker.ietf.org/doc/html/rfc8707#section-2.2) the access token request in the 'authorization_code' grant type should also contain the resource indicator.

### Steps To Reproduce

Example configuration:
```c#
builder.Services.AddAuthentication()
    .AddOpenIdConnect(options =>
    {
        options.Authority = "https://localhost:5001/";
        options.Resource = "urn:test";

        options.Scope.Add("profile");

        options.ClientId = "testclient";
        options.ClientSecret = "secret";
        options.ResponseType = "code";

        // workaround:
        options.Events = new OpenIdConnectEvents
        {
            OnAuthorizationCodeReceived = context =>
            {
                // the resource property here is null but should be set
                context.TokenEndpointRequest.Resource = "urn:test";

                return Task.FromResult(0);
            }
        }
    };
```

### .NET Version

6.0.201

### Anything else?

A fix may be applied somewhere around [this line](https://github.com/dotnet/aspnetcore/blob/c85baf8db0c72ae8e68643029d514b2e737c9fae/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L1110).

Something simple like this may already be enough:
```c#
if (Options.Resource != null)
{
    tokenEndpointRequest.Resource = Options.Resource;
}
```

**Comment by brockallen** [comment]
Just for context, the `Resource` setting in the current OIDC plumbing was added pre-RFC8707 and was for an AAD-specific feature, IIRC.

**Comment by Tratcher** [comment]
@jennyf19 Is the Resource supposed to be used in the access token request?

**Comment by jennyf19** [comment]
@Tratcher No, on the Azure AD v2 endpoint scopes are used, not resource. Resource was an AAD v1 concept which was not standard compliant.

**Comment by brockallen** [comment]
And in RFC8707 referenced above, multiple resource params are possible (and expected) on the authorize endpoint. And not a single param with multiple values mind you, but multiple instances of the same param each with one value. IIRC the current OIDC plumbing can't handle this as specified by the RFC.

**Comment by adityamandaleeka** [comment]
@blowdart Should we pursue this?

**Comment by blowdart** [comment]
Yes, but the question is do we reuse the existing parameter, or pull it out, and then replace it with a Resources one. Would that break existing users who never moved to Identity Web for aad?

**Comment by blowdart** [comment]
@jennyf19 @brentschmaltz Now that there's a real spec that clashes with AAD's custom use of resources IdentityModel needs correcting to match RFC8707. Brent have you looked into this? If not, can you?

**Comment by brentschmaltz** [comment]
@blowdart just read the spec, interesting that this is similar to [SignedHttpRequest](https://datatracker.ietf.org/doc/html/[draft-ietf-oauth-signed-http-request-03]) - draft, 'u' parameter.

This does map nicely to some work we are completing where inbound policy has an 'address' that could be used to identify policies that are applicable to a token.

What type of changes are you suggesting?

**Comment by adityamandaleeka** [comment]
@brentschmaltz Any further comment on this one?

**Comment by blowdart** [comment]
@jmprieur As Brent isn't answering, could you take a look here?

**Comment by jmprieur** [comment]
@blowdart : will discuss this with @brentschmaltz at the beginning of next week (beginning of August)
cc: @jennyf19

**Comment by jpda** [comment]
Any movement on this? We are designing our oauth2 systems and determining if we follow 8707 or mimic AAD's resource/scope parsing. A lot of our customers are dotnet shops and requiring a workaround if we adopt 8707 wouldn't be ideal.

**Comment by IceCoffee1024** [comment]
+1

**Comment by brentschmaltz** [comment]
@jpda where does the fix live?

**Comment by DeagleGross** [comment]
Thanks everyone for the continued interest — this needs a bit of design work (aligning the legacy single Resource property with RFC 8707's multi-resource model without breaking existing users), so we're moving it to .NET 12 Planning, and we'd happily review a proposal or PR that sketches out the approach
